### ISSUE TITLE
2D Physics performance issue

### ISSUE THREAD

**Issue by marynate** [godotengine/godot#600]
How to reproduce:
1. In 2D Platformer project, open bullet.xml
2. Select "Timer" node, change Wait Time to longer period, e.g. 300
3. Publish game to mobile device, shooting bullets

In iPad2:
    0 bullets  - 60fps
    10 bullets - 54
    20 bullets - 30 fps
    30 bullets - 20
    40 bullets - 20 
    50 bullets - 15

Found one blog post doing physics benchmark on iPad1 in 2011 (http://tech.flyclops.com/battle-of-the-ios-physics-engines-197), here's the their result of box2d on iPad 1:

```
200 rigidbody 60fps
370 rigidbody 30fps
406 rigidbody 15fps
```

**Comment by reduz** [comment]
i tested this a bit, and the main reason of the performance hit according to callgrind is that 2D math functions in Godot are not inlined. so project_range() and solve()

From the few things i saw in the callgraph, there really are many issues to optimize, but I would really rather a set of benchmarks is properly done and put together before this

**Comment by reduz** [comment]
I want to add that I'm not that experienced in SIMD optimization using autovectorization. I know that technically if your code emulates the way simd works, the autovectorizer will do it's job but my lack of experience with SIMD assembler (both SSE and NEON) would make it difficult to test if the compiler is optimizing properly.

**Comment by rraallvv** [comment]
I'would like to start making some tests with SIMD, pretty much it would be just copy-paste-ing the assembler code used for all the vectors and matrix algebra from other physics engines, mostly bullet physics. Godot already does almost all the memory allocations aligned, so it would be rather easy to just add the vectorization part.

I started making some research on how to do those optimizations relying on the compiler to do it's thing, but no luck, there should be a reason why nobody rely on the compiler to do those things, not even for copying memory.

[This is the scene](https://github.com/rraallvv/godot/tree/master/demos/performance/physics_3d) I'm using to do the profiling and testing.

![output](https://cloud.githubusercontent.com/assets/1840524/5890577/fc1f281c-a433-11e4-8696-17b2e75bb6c0.gif)

For instance that very same scene created "by hand" with bullet physics and GLKit, runs at 60 fps on an iPod Touch 5G, using NEON, [here is the repo](https://github.com/rraallvv/BulletPhysicsGLKit) if anyone wants to give it a try. Godot runs the same scene at 4~6 FPS. That's the reason why I'm eager to start doing optimizations.

Also I had to hack the classes `Vector` and `DVector`, because they reserve the first bytes for the ref counting and elements size, therefore misaligning the data. The class `Variant` has its data misaligned too, so I had to allocate the data dynamically.

I will post the PR if anyone wants to do some tests.

**Comment by akien-mga** [comment]
First of all thank you for your report and sorry for the delay.

We released Godot 3.0 in January 2018 after 18 months of work, fixing many old issues either directly, or by obsoleting/replacing the features they were referring to.

We still have hundreds of issues whose relevance/reproducibility needs to be checked against the [current stable version](https://godotengine.org/download), and that's where you can help us.
**Could you check if the issue that you described initially is still relevant/reproducible in Godot 3.0 or any newer version, and comment about it here?**

For bug reports, please also make sure that the issue contains detailed steps to reproduce the bug and, if possible, a zipped project that can be used to reproduce it right away. This greatly speeds up debugging and bugfixing tasks for our contributors.

Our Bugsquad will review this issue more in-depth in 15 days, and potentially close it if its relevance could not be confirmed.

Thanks in advance.

*Note: This message is being copy-pasted to many "stale" issues (90+ days without activity). It might happen that it is not meaningful for this specific issue or appears oblivious of the issue's context, if so please comment to notify the Bugsquad about it.*

**Comment by akien-mga** [comment]
As there was no answer to the above request, we're closing this issue as inactive.
If you can still reproduce it, please open a new issue against a currently supported release of Godot.
